File: FlickerNumber/Classes/UILabel+FlickerNumber.m
//
//  UILabel+FlickerNumber.m
//  FlickerNumber
//
//  Created by DeJohn Dong on 15-2-1.
//  Copyright (c) 2015年 DDKit. All rights reserved.
//

#import "UILabel+FlickerNumber.h"
#import <objc/runtime.h>

#define DDRangeIntegerKey @"RangeKey"
#define DDMultipleKey @"MultipleKey"
#define DDBeginNumberKey @"BeginNumberKey"
#define DDEndNumberKey @"EndNumberKey"
#define DDResultNumberKey @"ResultNumberKey"

#define DDAttributeKey @"AttributeKey"
#define DDFormatKey @"FormatStringKey"

#define DDFrequency 1.0/30.0f

#define DDDictArrtributeKey @"attribute"
#define DDDictRangeKey @"range"


@interface UILabel ()

@property (nonatomic, strong) NSNumber *flickerNumber;
@property (nonatomic, strong) NSNumberFormatter *flickerNumberFormatter;
@property (nonatomic, strong) NSTimer *currentTimer;

@end

@implementation UILabel (FlickerNumber)

#pragma mark - runtime methods

- (void)setFlickerNumber:(NSNumber *)flickerNumber{
    objc_setAssociatedObject(self, @selector(flickerNumber), flickerNumber, OBJC_ASSOCIATION_RETAIN_NONATOMIC);
}

- (NSNumber *)flickerNumber{
    return objc_getAssociatedObject(self, _cmd);
}

- (void)setFlickerNumberFormatter:(NSNumberFormatter *)flickerNumberFormatter{
    objc_setAssociatedObject(self, @selector(flickerNumberFormatter), flickerNumberFormatter, OBJC_ASSOCIATION_RETAIN_NONATOMIC);
}

- (NSNumberFormatter *)flickerNumberFormatter{
    return objc_getAssociatedObject(self, _cmd);
}

- (void)setCurrentTimer:(NSTimer *)currentTimer{
    objc_setAssociatedObject(self, @selector(currentTimer), currentTimer, OBJC_ASSOCIATION_RETAIN_NONATOMIC);
}

- (NSTimer *)currentTimer{
    return objc_getAssociatedObject(self, _cmd);
}

#pragma mark - flicker methods(public)

- (void)dd_setNumber:(NSNumber *)number{
    [self dd_setNumber:number duration:1.0 format:nil attributes:nil];
}

- (void)dd_setNumber:(NSNumber *)number formatter:(NSNumberFormatter *)formatter{
    if(!formatter)
        formatter = [self defaultFormatter];
    [self dd_setNumber:number duration:1.0 format:nil numberFormatter:formatter attributes:nil];
}

- (void)dd_setNumber:(NSNumber *)number format:(NSString *)formatStr{
    [self dd_setNumber:number duration:1.0 format:formatStr attributes:nil];
}

- (void)dd_setNumber:(NSNumber *)number format:(NSString *)formatStr formatter:(NSNumberFormatter *)formatter{
    if(!formatter)
        formatter = [self defaultFormatter];
    [self dd_setNumber:number duration:1.0 format:formatStr numberFormatter:formatter attributes:nil];
}

- (void)dd_setNumber:(NSNumber *)number attributes:(id)attrs{
    [self dd_setNumber:number duration:1.0 format:nil attributes:attrs];
}


- (void)dd_setNumber:(NSNumber *)number formatter:(NSNumberFormatter *)formatter attributes:(id)attrs{
    if(!formatter)
        formatter = [self defaultFormatter];
    [self dd_setNumber:number duration:1.0 format:nil numberFormatter:formatter attributes:attrs];
}

- (void)dd_setNumber:(NSNumber *)number duration:(NSTimeInterval)duration format:(NSString *)formatStr{
    [self dd_setNumber:number duration:duration format:formatStr attributes:nil];
}

- (void)dd_setNumber:(NSNumber *)number duration:(NSTimeInterval)duration format:(NSString *)formatStr numberFormatter:(NSNumberFormatter *)formatter{
    if(!formatter)
        formatter = [self defaultFormatter];
    [self dd_setNumber:number duration:duration format:formatStr numberFormatter:formatter attributes:nil];
}

- (void)dd_setNumber:(NSNumber *)number duration:(NSTimeInterval)duration{
    [self dd_setNumber:number duration:duration format:nil attributes:nil];
}

- (void)dd_setNumber:(NSNumber *)number duration:(NSTimeInterval)duration formatter:(NSNumberFormatter *)formatter{
    if(!formatter)
        formatter = [self defaultFormatter];
    [self dd_setNumber:number duration:duration format:nil numberFormatter:formatter attributes:nil];
}

- (void)dd_setNumber:(NSNumber *)number duration:(NSTimeInterval)duration attributes:(id)attrs{
    [self dd_setNumber:number duration:duration format:nil attributes:attrs];
}

- (void)dd_setNumber:(NSNumber *)number duration:(NSTimeInterval)duration format:(NSString *)formatStr formatter:(NSNumberFormatter *)formatter{
    if(!formatter)
        formatter = [self defaultFormatter];
    [self dd_setNumber:number duration:duration format:formatStr formatter:formatter];
}

- (void)dd_setNumber:(NSNumber *)number duration:(NSTimeInterval)duration formatter:(NSNumberFormatter *)formatter attributes:(id)attrs{
    if(!formatter)
        formatter = [self defaultFormatter];
    [self dd_setNumber:number duration:duration format:nil numberFormatter:formatter attributes:attrs];
}

- (void)dd_setNumber:(NSNumber *)number duration:(NSTimeInterval)duration format:(NSString *)formatStr attributes:(id)attrs{
    [self dd_setNumber:number duration:duration format:formatStr numberFormatter:nil attributes:attrs];
}

- (void)dd_setNumber:(NSNumber *)number duration:(NSTimeInterval)duration format:(NSString *)formatStr numberFormatter:(NSNumberFormatter *)formatter attributes:(id)attrs {
    /**
     *  check the number type
     */
    NSAssert([number isKindOfClass:[NSNumber class]], @"Number Type is not matched , exit");
    if(![number isKindOfClass:[NSNumber class]]) {
        self.text = [NSString stringWithFormat:@"%@",number];
        return;
    }
    
    /* limit duration is postive number and it is large than 0.3 , fixed the issue#1--https://github.com/openboy2012/FlickerNumber/issues/1 */
    duration = fabs(duration) < 0.3 ? 0.3 : fabs(duration);
    
    [self.currentTimer invalidate];
    self.currentTimer = nil;
    
    //initialize useinfo dict
    NSMutableDictionary *userInfo = [NSMutableDictionary dictionaryWithCapacity:0];
    
    if(formatStr)
        [userInfo setObject:formatStr forKey:DDFormatKey];
    
    [userInfo setObject:number forKey:DDResultNumberKey];
    
    //initialize variables
    long long beginNumber = 0;
    [userInfo setObject:@(beginNumber) forKey:DDBeginNumberKey];
    self.flickerNumber = @0;
    unsigned long long endNumber = [number unsignedLongLongValue];
    
    //get multiple if number is double type
    int multiple = [self multipleForNumber:number formatString:formatStr];
    if (multiple > 0)
        endNumber = [number doubleValue] * multiple;
    
    //check the number if out of bounds the unsigned int length
    if(endNumber >= INT64_MAX){
        self.text = [NSString stringWithFormat:@"%@",number];
        return;
    }
    
    [userInfo setObject:@(multiple) forKey:DDMultipleKey];
    [userInfo setObject:@(endNumber) forKey:DDEndNumberKey];
    if((endNumber * DDFrequency)/duration < 1){
        duration = duration * 0.3;
    }
    [userInfo setObject:@((endNumber * DDFrequency)/duration) forKey:DDRangeIntegerKey];
    
    if(attrs)
        [userInfo setObject:attrs forKey:DDAttributeKey];
    
    self.flickerNumberFormatter = nil;
    if(formatter)
        self.flickerNumberFormatter = formatter;
    
    self.currentTimer = [NSTimer scheduledTimerWithTimeInterval:DDFrequency target:self selector:@selector(flickerAnimation:) userInfo:userInfo repeats:YES];
    [[NSRunLoop currentRunLoop] addTimer:self.currentTimer forMode:NSRunLoopCommonModes];
}


#pragma mark - private methods
/**
 *  flicker number animation
 *
 *  @param timer schedule timer
 */
- (void)flickerAnimation:(NSTimer *)timer{
    /**
     *  check the rangeNumber if more than 1.0, fixed the issue#2--https://github.com/openboy2012/FlickerNumber/issues/2
     */
    if ([timer.userInfo[DDRangeIntegerKey] floatValue] >= 1.0) {
        long long rangeInteger = [timer.userInfo[DDRangeIntegerKey] longLongValue];
        self.flickerNumber = @([self.flickerNumber longLongValue] + rangeInteger);
    } else {
        float rangeInteger = [timer.userInfo[DDRangeIntegerKey] floatValue];
        self.flickerNumber = @([self.flickerNumber floatValue] + rangeInteger);
    }
    
    
    int multiple = [timer.userInfo[DDMultipleKey] intValue];
    if(multiple > 0) {
        [self floatNumberHandler:timer andMultiple:multiple];
    }else {
        NSString *formatStr = timer.userInfo[DDFormatKey]?:(self.flickerNumberFormatter?@"%@":@"%.0f");
        self.text = [self finalString:@([self.flickerNumber longLongValue]) stringFormat:formatStr numberFormatter:self.flickerNumberFormatter];
        
        if(timer.userInfo[DDAttributeKey]){
            [self attributedHandler:timer.userInfo[DDAttributeKey]];
        }
        
        if([self.flickerNumber longLongValue] >= [timer.userInfo[DDEndNumberKey] longLongValue]){
            self.text = [self finalString:timer.userInfo[DDResultNumberKey] stringFormat:formatStr numberFormatter:self.flickerNumberFormatter];
            if(timer.userInfo[DDAttributeKey]){
                [self attributedHandler:timer.userInfo[DDAttributeKey]];
            }
            [timer invalidate];
        }
    }
}

/**
 *  float number handler
 *
 *  @param timer    timer
 *  @param multiple multiple
 */
- (void)floatNumberHandler:(NSTimer *)timer andMultiple:(int)multiple {
    NSString *formatStr = timer.userInfo[DDFormatKey]?:(self.flickerNumberFormatter?@"%@":[NSString stringWithFormat:@"%%.%df",(int)log10(multiple)]);
    self.text = [self finalString:@([self.flickerNumber doubleValue]/multiple) stringFormat:formatStr numberFormatter:self.flickerNumberFormatter];
    if(timer.userInfo[DDAttributeKey]){
        [self attributedHandler:timer.userInfo[DDAttributeKey]];
    }
    if([self.flickerNumber longLongValue] >= [timer.userInfo[DDEndNumberKey] longLongValue]){
        self.text = [self finalString:timer.userInfo[DDResultNumberKey] stringFormat:formatStr numberFormatter:self.flickerNumberFormatter];
        if(timer.userInfo[DDAttributeKey]){
            [self attributedHandler:timer.userInfo[DDAttributeKey]];
        }
        [timer invalidate];
    }
}

/**
 *  attributes string handle methods
 *
 *  @param attributes attributes variable
 */
- (void)attributedHandler:(id)attributes {
    if ([attributes isKindOfClass:[NSDictionary class]]) {
        NSRange range = [attributes[DDDictRangeKey] rangeValue];
        [self addAttributes:attributes[DDDictArrtributeKey] range:range];
    } else if([attributes isKindOfClass:[NSArray class]]) {
        for (NSDictionary *attribute in attributes) {
            NSRange range = [attribute[DDDictRangeKey] rangeValue];
            [self addAttributes:attribute[DDDictArrtributeKey] range:range];
        }
    }
}

/**
 *  attributes string result methods
 *
 *  @param attri attribute
 *  @param range range
 */
- (void)addAttributes:(NSDictionary *)attri
                range:(NSRange)range {
    NSMutableAttributedString *str = [[NSMutableAttributedString alloc] initWithAttributedString:self.attributedText];
    // handler the out range exception
    if(range.location + range.length <= str.length){
        [str addAttributes:attri range:range];
    }
    self.attributedText = str;
}

/**
 *  get muliple from number
 *
 *  @param number past number
 *
 *  @return mulitple
 */
- (int)multipleForNumber:(NSNumber *)number
            formatString:(NSString *)formatStr {
    if([formatStr rangeOfString:@"%@"].location == NSNotFound) {
        formatStr = [self regexNumberFormat:formatStr];
        NSString *formatNumberString = [NSString stringWithFormat:formatStr,[number floatValue]];
        if([formatNumberString rangeOfString:@"."].location != NSNotFound){
            NSUInteger length = [[formatNumberString substringFromIndex:[formatNumberString rangeOfString:@"."].location +1] length];
            float padding = log10f(length<6? length:6);
            number = @([formatNumberString floatValue] + padding);
        }
    }
    
    NSString *str = [NSString stringWithFormat:@"%@",number];
    if([str rangeOfString:@"."].location != NSNotFound) {
        NSUInteger length = [[str substringFromIndex:[str rangeOfString:@"."].location +1] length];
        // Max Multiple is 6
        return  length >= 6 ? pow(10, 6): pow(10, (int)length);
    }
    return 0;
}

- (NSString *)stringFromNumber:(NSNumber *)number
               numberFormatter:(NSNumberFormatter *)formattor {
    if(!formattor) {
        formattor = [[NSNumberFormatter alloc] init];
        formattor.formatterBehavior = NSNumberFormatterBehavior10_4;
        formattor.numberStyle = NSNumberFormatterDecimalStyle;
    }
    return [formattor stringFromNumber:number];
}

- (NSString *)finalString:(NSNumber *)number
             stringFormat:(NSString *)formatStr
          numberFormatter:(NSNumberFormatter *)formatter {
    NSString *finalString = nil;
    if(formatter){
        finalString = [NSString stringWithFormat:formatStr,[self stringFromNumber:number numberFormatter:formatter]];
    }else{
        NSAssert([formatStr rangeOfString:@"%@"].location == NSNotFound, @"string format type is not matched,please check your format type");
        finalString = [NSString stringWithFormat:formatStr,[number doubleValue]];
    }
    return finalString;
}

/**
 *  get the decimal style number as default number formatter
 *
 *  @return number formatter
 */
- (NSNumberFormatter *)defaultFormatter {
    NSNumberFormatter *formattor = [[NSNumberFormatter alloc] init];
    formattor.formatterBehavior = NSNumberFormatterBehavior10_4;
    formattor.numberStyle = NSNumberFormatterDecimalStyle;
    return formattor;
}

/**
 *  get the format string use regex method
 *
 *  @param formatString origin string
 *
 *  @return format string
 */
- (NSString *)regexNumberFormat:(NSString *)formatString {
    NSError *regexError = nil;
    NSRegularExpression *regex =
    [NSRegularExpression regularExpressionWithPattern:@"^%((\\d+.\\d+)|(\\d+).|(.\\d+))f$"
                                              options:NSRegularExpressionCaseInsensitive
                                                error:&regexError];
    if (!regexError) {
        NSTextCheckingResult *match = [regex firstMatchInString:formatString
                                                        options:0
                                                          range:NSMakeRange(0, [formatString length])];
        if (match) {
            NSString *result = [formatString substringWithRange:match.range];
            return result;
        }
    } else {
        NSLog(@"error - %@", regexError);
    }
    return @"%f";
}

@end

@implementation NSDictionary(FlickerNumber)

+ (instancetype)dictionaryWithAttribute:(NSDictionary *)attribute range:(NSRange)range {
    NSMutableDictionary *dict = [NSMutableDictionary dictionaryWithCapacity:0];
    [dict setObject:attribute forKey:DDDictArrtributeKey];
    [dict setObject:[NSValue valueWithRange:range] forKey:DDDictRangeKey];
    return dict;
}

@end

File: FlickerNumber/Classes/UILabel+FlickerNumber.swift
//
//  UILabel+FlickerNumber.swift
//  FlickerNumber
//
//  Created by DeJohn Dong on 15/10/23.
//  Copyright © 2015年 ddkit. All rights reserved.
//

import Foundation
import UIKit

private let rangeIntegerName = "rangeInteger"
private let multipleName = "multiple"
private let beginNumberName = "beginNumber"
private let endNumberName = "endNumber"
private let resultNumberName = "resultNumber"
private let attributeName = "attribute"
private let formatName = "format"
private let frequency = 1.0/30.0

public extension UILabel {
    public func dd_setNumber(number : NSNumber) {
        self.dd_setNumber(number, duration: 1.0, format: nil, numberFomatter: nil, attributes: nil)
    }
    
    public func dd_setNumber(number : NSNumber, duration : NSTimeInterval, format : String?, numberFomatter : NSNumberFormatter?, attributes : AnyObject? ) {
        // check the number type
        assert(number.isKindOfClass(NSNumber), "Number Type is not matched, exit")
        if(!number.isKindOfClass(NSNumber)) {
            text = String(number)
            return
        }
        
        //limit duration is postive number and it is large than 0.3
        var durationTime : NSTimeInterval = fabs(duration) < 0.3 ? 0.3 : fabs(duration)
        
        self.fn_timer?.invalidate()

        //initialize useinfo dict
        let userDict : NSMutableDictionary = NSMutableDictionary()
        
        if format != nil {
            userDict.setValue(format, forKey:formatName)
        }
        
        userDict.setValue(number, forKey: resultNumberName)
        
        //initialize variables
        let beginNumber : NSNumber = 0
        userDict.setValue(beginNumber, forKey: beginNumberName)
        
        fn_setNumber(0)
        var endNumber : Int64 = number.longLongValue;

        //get multiple if number is float & double type
        let multiple = self.multipleForNumber(number, formatString: format)
        
        if multiple > 0 {
            endNumber = Int64(number.doubleValue * Double(multiple))
        }
        //check the number if out of bounds the unsigned int length
        if endNumber >= INT64_MAX {
            self.text = String(format:"%@", number);
            return;
        }
        
        userDict.setValue(multiple, forKey: multipleName)
        userDict.setValue(NSNumber(longLong: endNumber), forKey: endNumberName)
        
        if (Double(endNumber) * frequency)/durationTime < 1.0 {
            durationTime = durationTime * 0.3
        }
        let rangeNumber = (Double(endNumber) * frequency)/durationTime
        userDict.setValue(NSNumber(double: rangeNumber), forKey: rangeIntegerName)
        
        if attributes != nil {
            userDict.setValue(attributes, forKey: attributeName)
        }
        
        if numberFomatter != nil {
            fn_setFormatter(numberFomatter!)
        }
        
        fn_setTimer(NSTimer.scheduledTimerWithTimeInterval(frequency,
            target: self,
            selector: "flickerAnimation:",
            userInfo: userDict,
            repeats: true))
        
        NSRunLoop.currentRunLoop().addTimer(self.fn_timer!, forMode: NSRunLoopCommonModes)
    }
    
    public func flickerAnimation(timer : NSTimer) -> Void {
        let rangeInteger = timer.userInfo?.valueForKey(rangeIntegerName)?.longLongValue;
        let resultInteger = (self.fn_number?.longLongValue)! + rangeInteger!;
        fn_setNumber(NSNumber(longLong: resultInteger))
        
        let multiple = timer.userInfo?.valueForKey(multipleName)?.intValue
        
        if multiple > 0 {
            self.fn_finalNumberDoubleType(timer, multiple: Int(multiple!))
        } else {
            var formatStr : String = String()
            if timer.userInfo?.valueForKey(formatName) != nil {
                formatStr = timer.userInfo?.valueForKey(formatName) as! String
            }
            if formatStr.isEmpty {
                if self.fn_formatter != nil {
                    formatStr = "%@"
                } else {
                    formatStr = "%.0f"
                }
            }
            self.text = self.fn_finalNumber(self.fn_number!, format:formatStr, numberFormatter: self.fn_formatter)
            
            let attributes = timer.userInfo?.valueForKey(attributeName)
            
            if attributes != nil {
                self.drawAttributes(attributes!)
            }
            
            let endNumber : NSNumber = timer.userInfo?.valueForKey(endNumberName) as! NSNumber
            
            if self.fn_number?.longLongValue >= endNumber.longLongValue {
                let resultNumber : NSNumber = timer.userInfo?.valueForKey(resultNumberName) as! NSNumber
                self.text = self.fn_finalNumber(resultNumber, format:formatStr, numberFormatter: self.fn_formatter)
                
                let attributes = timer.userInfo?.valueForKey(attributeName)
                
                if attributes != nil {
                    self.drawAttributes(attributes!)
                }
                self.fn_timer?.invalidate()
            }
        }
    }
    
    private func fn_finalNumberDoubleType(timer : NSTimer, multiple: Int) {
        var formatStr : String = String()
        if timer.userInfo?.valueForKey(formatName) != nil {
            formatStr = timer.userInfo?.valueForKey(formatName) as! String
        }
        if formatStr.isEmpty {
            if self.fn_formatter != nil {
                formatStr = "%@"
            } else {
                formatStr = String(format: "%%.%df", Int(log10(Float(multiple))))
            }
        }
        let flickerNumber = (self.fn_number?.doubleValue)! / Double(multiple)
        
        self.text = self.fn_finalNumber(flickerNumber, format:formatStr, numberFormatter: self.fn_formatter)
        
        let attributes = timer.userInfo?.valueForKey(attributeName)
        
        if attributes != nil {
            self.drawAttributes(attributes!)
        }
        
        let endNumber : NSNumber = timer.userInfo?.valueForKey(endNumberName) as! NSNumber
        
        if self.fn_number?.longLongValue >= endNumber.longLongValue {
            let resultNumber : NSNumber = timer.userInfo?.valueForKey(resultNumberName) as! NSNumber
            
            self.text = self.fn_finalNumber(resultNumber, format:formatStr, numberFormatter: self.fn_formatter)
            
            let attributes = timer.userInfo?.valueForKey(attributeName)
            
            if attributes != nil {
                self.drawAttributes(attributes!)
            }
            self.fn_timer?.invalidate()
        }
    }
    
    private func fn_stringFromNumber(number : NSNumber, numberFormatter: NSNumberFormatter) -> String? {
           return numberFormatter.stringFromNumber(number)
    }
    
    private func fn_finalNumber(number : NSNumber, format: String, numberFormatter: NSNumberFormatter?) -> String? {
        var finalString = String()
        if numberFormatter != nil {
            finalString = String(format: format, self.fn_stringFromNumber(number, numberFormatter: numberFormatter!)!)
        } else {
            assert(format.rangeOfString("%@") == nil, "string format type is not matched, please check your format type.")
            finalString = String(format: format, number.doubleValue)
        }
        return finalString;
    }
    
    private func drawAttributes(attributes : AnyObject) -> Void {
        if attributes.isKindOfClass(NSDictionary) {
            let range = attributes.valueForKey(fnRangeName)?.rangeValue
            let attribute = attributes.valueForKey(fnAttributeName)
            self.addAttributes(attribute!, range : range!)
        } else if attributes.isKindOfClass(NSArray) {
            for attribute in attributes as! NSArray {
                let range = attribute.valueForKey(fnRangeName)?.rangeValue
                let attri = attribute.valueForKey(fnAttributeName)
                self.addAttributes(attri!, range : range!)
            }
        }
    }
    
    private func addAttributes(attributes : AnyObject, range : NSRange) -> Void {
        let attributedStr = NSMutableAttributedString(attributedString : self.attributedText!)
        if range.location + range.length < attributedStr.length  {
            attributedStr.addAttributes(attributes as! [String : AnyObject], range: range)
        }
        self.attributedText = attributedStr;
    }
    
    private func multipleForNumber(number : NSNumber, formatString : String?) -> Int {
        var newNumber = number;
        if formatString != nil && formatString!.rangeOfString("%@") == nil {
            let fStr : String = self.regexNumberFormat(formatString!)
            let formatNumberStr : String = String(fStr, number.floatValue)
            if formatNumberStr.rangeOfString(".") != nil {
                let fn_range = formatNumberStr.rangeOfString(".")
                let index = fn_range?.startIndex.advancedBy(1)
                let subStr : String = formatNumberStr.substringFromIndex(index!)
                let length : Int = subStr.characters.count < 6 ? subStr.characters.count : 6
                let padding = log10f(Float(length))
                newNumber = (formatNumberStr as NSString).floatValue + padding
            }
        }
        
        let resultStr = String(format: "%@", newNumber)
        if resultStr.rangeOfString(".") != nil {
            let re_range = resultStr.rangeOfString(".")
            let index = re_range?.startIndex.advancedBy(1)
            let subStr : String = resultStr.substringFromIndex(index!)
            // Max Multiple is 6
            let length : Int = subStr.characters.count >= 6 ? 6 : subStr.characters.count
            return Int(powf(10, Float(length)))
        }
        return 0
    }
    
    private func regexNumberFormat(format : String) -> String {
        let patternStr : String = "^%((\\d+.\\d+)|(\\d+).|(.\\d+))f$"
        do {
            let regex : NSRegularExpression =  try NSRegularExpression(pattern: patternStr, options: NSRegularExpressionOptions.CaseInsensitive)
            let fRange = NSMakeRange(0, format.characters.count)
            let match : NSTextCheckingResult = regex.firstMatchInString(format, options: NSMatchingOptions.ReportProgress, range: fRange)!
            let result : NSString = (format as NSString).substringWithRange(match.range);
            return result as String;
        } catch {
        }
        return "%f"
    }
}

private var flickerNumberKey : Void?
private var flickerNumberFormatterKey : Void?
private var flickerTimerKey : Void?

public extension UILabel {
    public var fn_number : NSNumber? {
        get {
            return objc_getAssociatedObject(self, &flickerNumberKey) as? NSNumber
        }
    }
    
    private func fn_setNumber(number : NSNumber) {
        objc_setAssociatedObject(self, &flickerNumberKey, number, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
    }
    
    public var fn_formatter : NSNumberFormatter? {
        get {
            return objc_getAssociatedObject(self, &flickerNumberFormatterKey) as? NSNumberFormatter
        }
    }
    
    private func fn_setFormatter(formatter : NSNumberFormatter) {
        objc_setAssociatedObject(self, &flickerNumberFormatterKey, formatter, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
    }
    
    public var fn_timer : NSTimer? {
        get {
            return objc_getAssociatedObject(self, &flickerTimerKey) as? NSTimer
        }
    }
    
    private func fn_setTimer(timer : NSTimer) {
        objc_setAssociatedObject(self, &flickerTimerKey, timer, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
    }
}

public let fnAttributeName = "attributeKey"
public let fnRangeName = "rangeKey"

public extension NSDictionary {
    public static func fn_dictionary(attribute : NSDictionary, range : NSRange) -> NSDictionary {
        let dictionary = NSMutableDictionary()
        dictionary.setObject(attribute , forKey: fnAttributeName)
        let rangeValue = NSValue(range: range)
        dictionary.setObject(rangeValue, forKey: fnRangeName)
        return dictionary
    }
}





